File: python/src/threads/dump_thread.py
import datetime
from PyQt5 import QtCore
from PyQt5.QtCore import QObject, pyqtProperty, pyqtSignal, pyqtSlot

from src.app.canbus import CanBus


class DumpThread(QObject):
    dumpSig = pyqtSignal(str, str, str, str, arguments=[
                         'time', 'can_id', 'dlc', 'data'])
    dumpDone = pyqtSignal()

    def __init__(self, parent=None):
        super(DumpThread, self).__init__(parent)

        self.can_bus = CanBus()
        self.__abort = False

    @pyqtSlot()
    def dump(self):
        while True:
            if self.__abort:
                break

            msg = self.can_bus.recv(0.5)
            if msg is not None:
                timestamp = msg.timestamp
                time = datetime.datetime.fromtimestamp(timestamp).strftime('%H:%M:%S')
                can_id = hex(msg.arbitration_id)
                dlc = str(msg.dlc).zfill(2)
                data = ' '.join(format(byte, 'x').zfill(2).upper()
                                for byte in msg.data)
                self.dumpSig.emit(time, can_id, dlc, data)

        self.dumpDone.emit()

    @pyqtSlot()
    def abort_dump(self):
        self.__abort = True

File: cpp/src/threads/dump_thread.cpp
#include <QString>

#include "dump_thread.h"
#include "src/helper.h"

DumpThread::DumpThread()
{
  can_bus_ = new CanBus();
  can_bus_->open();
}

void DumpThread::dump()
{
  QString time, can_id, dlc, data_string;

  while (true) {
    if (this->__abort) {
      break;
    }

    can_bus_->recv(msg_);

    time = QString::fromStdString(currentDateTime());
    can_id = QString::fromStdString(int_to_hex_prefix(int(msg_.can_id)));
    dlc = QString::fromStdString(zfill(int(msg_.can_dlc), 2));
    for (uint8_t i = 0; i < msg_.can_dlc; i++) {
      QString d = QString::fromStdString(int_to_hex(int(msg_.data[i]))).toUpper();
      data_string.append(d);
      data_string.append(" ");
    }
    emit dumpSig(time, can_id, dlc, data_string);
    data_string = "";
  }

  emit dumpDone();
}


void DumpThread::abortDump()
{
  this->__abort = true;
}
